src/main/java/org/junit/runner/manipulation/Ordering.java

```diff
@@ -0,0 +1,158 @@
+package org.junit.runner.manipulation;
+
+import java.lang.reflect.Constructor;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+
+import org.junit.runner.Description;
+import org.junit.runner.OrderWith;
+
+/**
+ * Reorders tests. An {@code Ordering} can reverse the order of tests, sort the
+ * order or even shuffle the order.
+ * 
+ * <p>In general you will not need to use a <code>Ordering</code> directly.
+ * Instead, use {@link org.junit.runner.Request#orderWith(Ordering)}.
+ *
+ * @since 4.13
+ */
+public abstract class Ordering {
+    private static final String CONSTRUCTOR_ERROR_FORMAT
+            = "Ordering class %s should have a public constructor with signature %s(Ordering.Context context)";
```

Comment:
I prefer not to for a few reasons:

1. It would make the API confusing, since context contains data about the top-level class that has the annotation, while the other parameter is about the current set of tests that need to be ordered.
2. The Ordering might need to do expensive initialization based on the content (David's example was loading the result of the previous run so failing tests run first). If we pass context via the method then either the implementation has to do that initialization each time or it has to do lazy initialization

Note we have similar requirements on Runner implementations